nack/receiver_interceptor.go

```diff
@@ -0,0 +1,174 @@
+package nack
+
+import (
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/pion/interceptor"
+	"github.com/pion/logging"
+	"github.com/pion/rtcp"
+	"github.com/pion/rtp"
+)
+
+// ReceiverInterceptor interceptor generates nack messages.
+type ReceiverInterceptor struct {
+	interceptor.NoOp
+	size        uint16
+	skipLastN   uint16
+	interval    time.Duration
+	receiveLogs *sync.Map
+	m           sync.Mutex
+	wg          sync.WaitGroup
+	close       chan struct{}
+	log         logging.LeveledLogger
+}
+
+// NewReceiverInterceptor returns a new ReceiverInterceptor interceptor
+func NewReceiverInterceptor(size uint16, skipLastN uint16, interval time.Duration, log logging.LeveledLogger) (*ReceiverInterceptor, error) {
+	_, err := NewReceiveLog(size)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ReceiverInterceptor{
+		NoOp:        interceptor.NoOp{},
+		size:        size,
+		skipLastN:   skipLastN,
+		interval:    interval,
+		receiveLogs: &sync.Map{},
+		close:       make(chan struct{}),
+		log:         log,
+	}, nil
+}
+
+// BindRTCPWriter lets you modify any outgoing RTCP packets. It is called once per PeerConnection. The returned method
+// will be called once per packet batch.
+func (n *ReceiverInterceptor) BindRTCPWriter(writer interceptor.RTCPWriter) interceptor.RTCPWriter {
+	n.m.Lock()
+	select {
+	case <-n.close:
+		// already closed
+		n.m.Unlock()
+		return writer
+	default:
+	}
+
+	n.wg.Add(1)
+	n.m.Unlock()
+
+	go n.loop(writer)
+
+	return writer
+}
+
+// BindRemoteStream lets you modify any incoming RTP packets. It is called once for per RemoteStream. The returned method
+// will be called once per rtp packet.
+func (n *ReceiverInterceptor) BindRemoteStream(info *interceptor.StreamInfo, reader interceptor.RTPReader) interceptor.RTPReader {
+	hasNack := false
+	for _, fb := range info.RTCPFeedback {
+		if fb.Type == "nack" && fb.Parameter == "" {
+			hasNack = true
+		}
+	}
+
+	if !hasNack {
+		return reader
+	}
+
+	// error is already checked in NewReceiverInterceptor
+	receiveLog, _ := NewReceiveLog(n.size)
+	n.receiveLogs.Store(info.SSRC, receiveLog)
+
+	return interceptor.RTPReaderFunc(func() (*rtp.Packet, interceptor.Attributes, error) {
+		p, attr, err := reader.Read()
+		if err != nil {
+			return nil, nil, err
+		}
+
+		receiveLog.Add(p.SequenceNumber)
+
+		return p, attr, nil
+	})
+}
+
+// UnbindLocalStream is called when the Stream is removed. It can be used to clean up any data related to that track.
+func (n *ReceiverInterceptor) UnbindLocalStream(info *interceptor.StreamInfo) {
+	n.receiveLogs.Delete(info.SSRC)
+}
+
+func (n *ReceiverInterceptor) Close() error {
+	defer n.wg.Wait()
+	n.m.Lock()
+	defer n.m.Unlock()
+
+	select {
+	case <-n.close:
+		// already closed
+		return nil
+	default:
+	}
+
+	close(n.close)
+
+	return nil
+}
+
+func (n *ReceiverInterceptor) loop(rtcpWriter interceptor.RTCPWriter) {
+	defer n.wg.Done()
+
+	senderSSRC := rand.Uint32()
+
+	ticker := time.NewTicker(n.interval)
```

Comment:
Rather than doing this at set intervals, wouldn't it be better to trigger when packet n is missing but packet n + k has been received, for some suitable value of k?  In the Unnamed SFU, k is computes as: at least 3 (since that is what TCP assumes, and anything that makes stronger assumptions than TCP is likely to break in the Internet), and no more than the tolerable latency divided by the packet rate.

Reply:
I think this can be improved in a later mr as this is an internal behaviour. Also there should be a limit to how many times a nack for a packet is requested. k is already configurable, it's the skipLastN parameter.

Reply:
Created ticket!

I bet there are a lot of things we can do to make this better. I am really happy with what we have today though :)